Implement `clip-distances` extension for GL and Vulkan backends

dzamkov opened the pull request:
**Connections**
#6236: Partially resolves

**Description**
Implements the [`clip-distances`](https://gpuweb.github.io/gpuweb/#dom-gpufeaturename-clip-distances) extension for GL and Vulkan backend. This was already implemented in Naga for the most part, so we just need to hook it up in the adapters. The GL backend is a bit tricky because we need to manually enable each clip plane we're using with the GL API.

**Testing**
- Added Naga snapshot tests
- Added GPU test
- Integrated into personal project and verified clip planes work as expected

**Squash or Rebase?**

Squash

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry.

dzamkov committed:
Basic implementation of `clip_distances` for Vulkan and GL backends

dzamkov committed:
Added GPU test for `clip-distances`

dzamkov committed:
Update feature array size

dzamkov committed:
Add changelog entry

teoxoy reviewed (commented):
Looks good overall!

Could you also add the validation for the enable extension and make sure there are at most 8 clip planes (make sure to allow `NonConstArrayLength` if `overrides_resolved` is false)? Pointers:

https://github.com/gfx-rs/wgpu/blob/d09c730c02e5449b2075e88490f7c66b727157d9/naga/src/front/wgsl/parse/mod.rs#L223-L231

https://github.com/gfx-rs/wgpu/blob/dcada3d858844697c1e3ab0865710fd2871a0583/naga/src/valid/interface.rs#L215

teoxoy commented:
Based on these findings https://github.com/gpuweb/gpuweb/issues/390#issuecomment-1956104959, it seems we could easily get this working on Metal as well just by advertising the feature in the backend.

Getting it to work on D3D12 is a bit trickier since we need to emit and write to both `SV_ClipDistance0` and `SV_ClipDistance1` if there are more than 4 clip planes.

dzamkov committed:
Validate `clip_distances` array size

dzamkov commented:
Looks like the Metal integration will be a bit [more complicated](https://github.com/gfx-rs/wgpu/actions/runs/15550026829/job/43778728987?pr=7730) than just enabling the feature. I don't have the tools, hardware or expertise for Metal development, so I'll probably have to leave that out of the PR.

teoxoy commented:
I see, that's ok we can leave Metal out for now then - thanks for trying it out!

dzamkov committed:
Check for `clip_distances` enable directive

dzamkov committed:
Consolidate code for generating `enable` directives in WGSL backend and add `clip_distances`.

dzamkov commented:
- Reverted Metal implementation attempt

- Added validation for enable directive

- Ensured enable directive is used when generating WGSL

- Added validation for clip distances array length

- Using a pipeline override as the number of clip distances is invalid, and this is already caught by [more general validation logic](https://github.com/gfx-rs/wgpu/blob/620c9d1e8bcdede08a76134692babd6ccc004ecb/naga/src/valid/type.rs#L614). According to the WGSL specs:

  > The only valid use of a fixed-size array with an element count that is an [override-expression](https://www.w3.org/TR/WGSL/#override-expressions) that is not a [const-expression](https://www.w3.org/TR/WGSL/#const-expressions) is as a [memory view](https://www.w3.org/TR/WGSL/#memory-view) in the [workgroup](https://www.w3.org/TR/WGSL/#address-spaces-workgroup) address space. This includes the [store type](https://www.w3.org/TR/WGSL/#store-type) of a workgroup variable.

teoxoy reviewed (approved):
Looks great, thanks!
